### PR TITLE
Capabilities policy

### PR DESCRIPTION
**Pull request by SeanTAllen**

### CODE REVIEW

File: internal/tools/securitypolicy/README.md [hunk 1]
```diff
@@ -210,42 +217,58 @@ Is the following Rego policy:
...
         "layers": ["fe84c9d5bfddd07a2624d00333cf13c1a9c941f3a261f13ead44fc6a93bc0e7a","4dedae42847c704da891a28c25d32201a1ae440bce2aecccfa8e6f03b97a6a6c","41d64cdeb347bf236b4c13b7403b633ff11f1cf94dbc7cf881a44d6da88c5156","eb36921e1f82af46dfe248ef8f1b3afb6a5230a64181d960d10237a08cd73c79","e769d7487cc314d3ee748a4440805317c19262c7acd2fdbdb0d47d2e4613a15c","1b80f120dbd88e4355d6241b519c3e25290215c469516b49dece9cf07175a766"],
-        "mounts": [{"destination": "/container/path/one", "options": ["rbind","rshared","rw"], "source": "sandbox:///host/path/one", "type": "bind"},{"destination": "/container/path/two", "options": ["rbind","rshared","ro"], "source": "sandbox://host/path/two", "type": "bind"}],
+        "mounts": [{"destination": "/container/path/one", "options": ["rbind","rshared","rw"], "source": "sandbox:///host/path/one", "type": "bind"},{"destination": "/container/path/two", "options": ["rbind","rshared","ro"], "source": "sandbox:///host/path/two", "type": "bind"}],
```
Comment (SeanTAllen) [inline internal/tools/securitypolicy/README.md:231]: @matajoh i copied the output here from an actual run, do we have a nug with this additional "/" for sandbox? should i go looking for that and open an independent PR to fix?

File: pkg/securitypolicy/regopolicy_test.go [hunk 2]
```diff
@@ -4262,15 +4490,18 @@ func setupRegoCreateContainerTest(gc *generatedConstraints, testContainer *secur
...
-		mounts:      copyMounts(mountSpec.Mounts),
-		policy:      policy,
+		envList:      copyStrings(envList),
+		argList:      copyStrings(testContainer.Command),
+		workingDir:   testContainer.WorkingDir,
+		containerID:  containerID,
+		sandboxID:    sandboxID,
+		mounts:       copyMounts(mountSpec.Mounts),
+		policy:       policy,
+		capabilities: &capabilities,
```
Comment (SeanTAllen) [inline pkg/securitypolicy/regopolicy_test.go:4504]: this should be a copy.

File: pkg/securitypolicy/regopolicy_test.go [hunk 3]
```diff
@@ -4754,13 +4992,14 @@ func setupRegoDropEnvsTest(disjoint bool) (*regoContainerTestConfig, error) {
...
-		mounts:      copyMounts(mountSpec.Mounts),
-		policy:      policy,
+		envList:      copyStrings(envList),
+		argList:      copyStrings(containers[2].Command),
+		workingDir:   containers[2].WorkingDir,
+		containerID:  containerIDs[2],
+		sandboxID:    sandboxID,
+		mounts:       copyMounts(mountSpec.Mounts),
+		policy:       policy,
+		capabilities: (*oci.LinuxCapabilities)(&containers[2].Capabilities),
```
Comment (SeanTAllen) [inline pkg/securitypolicy/regopolicy_test.go:5002]: this should be doing a toExternal() and a copy,

File: pkg/securitypolicy/framework.rego [hunk 4]
```diff
@@ -237,6 +237,25 @@ container_started {
     data.metadata.started[input.containerID]
 }
 
+capsList_ok(allowedCapList, capList) {
+    allowedCapList == ["ALL"]
+}
+
+capsList_ok(allowedCapList, capList) {
+    every cap in capList {
+        some allowed in allowedCapList
```
Comment (matajoh) [inline pkg/securitypolicy/framework.rego:246]: I think you can simplify this to:

```
allowedCapList[cap]
```
see https://www.openpolicyagent.org/docs/latest/policy-reference/#sets

File: pkg/securitypolicy/regopolicy_test.go [hunk 5]
```diff
@@ -983,6 +984,148 @@ func Test_Rego_EnforceCreateContainer_Same_Container_Twice(t *testing.T) {
...
+func Test_Rego_EnforceCreateContainer_Capabilities_NoMatches(t *testing.T) {
+	f := func(p *generatedConstraints) bool {
+		tc, err := setupSimpleRegoCreateContainerTest(p)
+		if err != nil {
+			t.Error(err)
+			return false
+		}
+
+		capabilities := tc.capabilities
+		capabilities.Bounding = alterCapabilitySet(testRand, capabilities.Bounding)
```
Comment (matajoh) [inline pkg/securitypolicy/regopolicy_test.go:996]: Nit: is it worth testing these separately, perhaps choosing which one to alter as part of the fuzzing?

File: pkg/securitypolicy/regopolicy_test.go [hunk 6]
```diff
@@ -983,6 +984,148 @@ func Test_Rego_EnforceCreateContainer_Same_Container_Twice(t *testing.T) {
...
+func Test_Rego_EnforceCreateContainer_Capabilities_SubsetMatches(t *testing.T) {
+	f := func(p *generatedConstraints) bool {
+		tc, err := setupSimpleRegoCreateContainerTest(p)
+		if err != nil {
+			t.Error(err)
+			return false
+		}
+
+		capabilities := tc.capabilities
+		capabilities.Bounding = subsetCapabilitySet(testRand, capabilities.Bounding)
```
Comment (matajoh) [inline pkg/securitypolicy/regopolicy_test.go:1026]: Is it worth only altering one of these at a time?

File: pkg/securitypolicy/regopolicy_test.go [hunk 7]
```diff
@@ -983,6 +984,148 @@ func Test_Rego_EnforceCreateContainer_Same_Container_Twice(t *testing.T) {
...
+func Test_Rego_EnforceCreateContainer_Capabilities_SupersetDoesntMatch(t *testing.T) {
+	f := func(p *generatedConstraints) bool {
+		tc, err := setupSimpleRegoCreateContainerTest(p)
+		if err != nil {
+			t.Error(err)
+			return false
+		}
+
+		capabilities := tc.capabilities
+		capabilities.Bounding = superCapabilitySet(testRand, capabilities.Bounding)
```
Comment (matajoh) [inline pkg/securitypolicy/regopolicy_test.go:1083]: Worth testing these one at a time instead of all together, perhaps choosing one randomly per test?

**Review by matajoh — COMMENTED** [review]
Just a few minor things, otherwise LGTM

File: pkg/securitypolicy/regopolicy_test.go (hunk 5 above)
Comment (SeanTAllen) [reply pkg/securitypolicy/regopolicy_test.go:996]: i considered that and i decided no. It won't prevent bugs but would make it slightly more obvious where a bug might be but at the expense of additional code.

File: pkg/securitypolicy/regopolicy_test.go (hunk 6 above)
Comment (SeanTAllen) [reply pkg/securitypolicy/regopolicy_test.go:1026]: see above.

File: pkg/securitypolicy/regopolicy_test.go (hunk 7 above)
Comment (SeanTAllen) [reply pkg/securitypolicy/regopolicy_test.go:1083]: see above.

Comment (SeanTAllen) [reply pkg/securitypolicy/regopolicy_test.go:1083]: yes should do this

File: pkg/securitypolicy/regopolicy_test.go (hunk 6 above)
Comment (SeanTAllen) [reply pkg/securitypolicy/regopolicy_test.go:1026]: no here

File: pkg/securitypolicy/framework.rego (hunk 4 above)
Comment (SeanTAllen) [reply pkg/securitypolicy/framework.rego]: done

File: pkg/securitypolicy/regopolicy_test.go (hunk 5 above)
Comment (SeanTAllen) [reply pkg/securitypolicy/regopolicy_test.go:996]: not this one

File: pkg/securitypolicy/regopolicy_test.go (hunk 7 above)
Comment (SeanTAllen) [reply pkg/securitypolicy/regopolicy_test.go]: done
